### PR TITLE
Webconfig

### PR DESCRIPTION
**Pull request by scopatz**
This is an initial, fully functioning webconfig.  This builds off of #3226.

It is implemented in elm (because that is the only JavaScript 'framework' I half-way know, and also it is awesome) with a bootstrap theme (becaue I wanted to avoid doing a lot of CSS work).  Happy to walk anyone through these bits, if needed!

Also, it currently only implements selecting prompts, the color style, and xontribs.  Happy to have other suggestions, but figured that a lot of improvements & extensions can be done as a separate PR.

### CODE REVIEW

**Comment by scopatz** [comment]
@xonsh/xore - I think this is ready for review! Also, sorry for dropping *another* huge PR on you...

**Comment by melund** [comment]
Funny coincidence. I stumbled on elm a few weeks ago listening to a webcast. I nearly got nerd sniped then. Elm looked really cool but it is a bit more than a JavaScript framework.

Do we need anything special to get started?

**Comment by scopatz** [comment]
> Funny coincidence. I stumbled on elm a few weeks ago listening to a webcast. I nearly got nerd sniped then.

hahah, amazing!

> Elm looked really cool but it is a bit more than a JavaScript framework.

Yeah it is! It is great!

> Do we need anything special to get started?

Not really, just install elm (which is on conda-forge, if that helps).  Also, if you want to minify it, use uglify-js (also on conda-forge).  There is the `elm-compile.xsh` script that I wrote and put in `xonsh/webconfig` (which needs to be run from that dir) to build the app.

From there `xonsh/webconfig`, you can run `python main.py` to launch the app, or run `xonfig web` from anywhere else.

**Comment by melund** [comment]
Thanks, I run into the following: 

```
    raise subprocess.CalledProcessError(rtn, spec.args, output=self.output)
subprocess.CalledProcessError: Command '['rst2html5.py', '--template', 'C:\\Users\\mel\\AppData\\Local\\Temp\\tmp1qaptrz1', 'C:\\Users\\mel\\AppData\\Local\\Temp\\tmpnsrnl_uq']' returned non-zero exit status 1.
```

Are you using something which is not in repository?

**Comment by scopatz** [comment]
Oh yeah, that comes with `docutils`, which is a dependency of `sphinx`.

**Comment by melund** [comment]
Ahh. I was using elm in a separate conda environment. But there is still some issues. On my Windows machine I can't start scripts directly like that. 

```
$ rst2html5.py
Traceback (Most recent call first):
OSError: [WinError 193] %1 is not a valid Win32 application
```
The following works though: 
```
$ python $(which rst2html.py)
```
but it still fails with: 
```
  InputError: [Errno 13] Permission denied: 'C:\\Users\\mel\\AppData\\Local\\Temp\\tmpn_3eucdi'
```

The problem is the way [you use NamedTemporaryFile](https://github.com/xonsh/xonsh/pull/3370/files#diff-d1ff63db5fc7810ccf8f7e95581c8934R65-R70)

It works differently between windows and unix. 

From the Python docs:
> Whether the name can be used to open the file a second time, while the named temporary file is still open, varies across platforms (it can be so used on Unix; it cannot on Windows NT or later).

The most portable option is to create a tempfile.TemporaryDirectory context manager, and just have normal files in that folder.

**Comment by mitnk** [comment]
A bit of testing from a Mac user :)

Before:

image

After `Colors` change after using `xonfig web`:

image

The `import` line is what I'm changing to, works fine. But I don't know why `ls` line missed colors now.

**Prompt** change in web page works but it seems re-changed my `Colors` config (I have changed it to orange, remember?), but `ls` colors are back.

image

**Comment by gforsyth** [comment]
I'm good with merging this and fixing things up later -- thoughts from @xonsh/xore ?  

Things I think this feature needs before we cut a release:
* [ ] Helpful error messages if you run `xonfig web` and you are missing any required deps
* [ ] Fix for the `NamedTemporaryFile` error that @melund mentioned

**Comment by Carreau** [comment]
I would also suggest to rename to `elm-xompile.xsh` :-P. 
Joke appart; looks great !

File: xonsh/xonfig.py [hunk 1]
```diff
@@ -740,10 +757,10 @@ def TAGLINES():
     ("{{INTENSE_BLACK}}", "<", "-"),
     "{{GREEN}}xonfig{{NO_COLOR}} tutorial    {{INTENSE_WHITE}}->    Launch the tutorial in "
     "the browser{{NO_COLOR}}",
-    "{{GREEN}}xonfig{{NO_COLOR}} wizard      {{INTENSE_WHITE}}->    Run the configuration "
-    "wizard and claim your shell {{NO_COLOR}}",
-    "{{INTENSE_BLACK}}(Note: Run the Wizard or create a {{RED}}~/.xonshrc{{INTENSE_BLACK}} file "
-    "to suppress the welcome screen)",
+    "{{GREEN}}xonfig{{NO_COLOR}} web         {{INTENSE_WHITE}}->    Run the configuration "
+    "tool in the browser and claim your shell {{NO_COLOR}}",
```
Comment (Carreau) [inline xonsh/xonfig.py]: Also, you have an extra `i` in claim. I feel like "Clam your Shell" is a good pun.

File: xonsh/webconfig/elm-src/App.elm [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+centeredDeck cards =
+    Html.div
+        [ class "card-deck justify-content-center" ]
+        (List.map Card.view cards)
+
+xontribCard : Model -> XontribData -> Card.Config Msg
+xontribCard model xd =
+    Card.config [ Card.attrs
+                    [ style "min-width" "20em"
+                    , style "max-width" "20em"
```
Comment (Carreau) [inline xonsh/webconfig/elm-src/App.elm]: I feel like 20em is not enough. But this is personal taste.

**Comment by melund** [comment]
> I'm good with merging this and fixing things up later -- thoughts from @xonsh/xore ?

👍 I can try to tackle the windows related problems later. 

I ran into a encoding issue. It is an easy fixed by not assuming utf-8 is default on all platforms. So always do `f.read(encoding="utf-8")` if you want it to run on Windows. 

```
Web config started at 'http://localhost:8421'. Hit Crtl+C to stop.
127.0.0.1 - - [28/Oct/2019 08:31:22] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [28/Oct/2019 08:31:22] "GET /js/bootstrap.min.css HTTP/1.1" 200 -
127.0.0.1 - - [28/Oct/2019 08:31:23] "GET /js/app.min.js HTTP/1.1" 200 -
127.0.0.1 - - [28/Oct/2019 08:31:25] code 404, message File not found
127.0.0.1 - - [28/Oct/2019 08:31:25] "GET /favicon.ico HTTP/1.1" 404 -
127.0.0.1 - - [28/Oct/2019 08:32:48] "POST /save HTTP/1.1" 200 -
Web Config Values:
{'colors': 'abap',
 'prompt': '{BOLD_GREEN}{user}@{hostname}{BOLD_BLUE} {cwd}{NO_COLOR}> ',
 'xontribs': []}
----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 58536)
Traceback (most recent call last):
  File "C:\Users\mel\Anaconda3\envs\elm\lib\socketserver.py", line 316, in _handle_request_noblock
    self.process_request(request, client_address)
  File "C:\Users\mel\Anaconda3\envs\elm\lib\socketserver.py", line 347, in process_request
    self.finish_request(request, client_address)
  File "C:\Users\mel\Anaconda3\envs\elm\lib\socketserver.py", line 360, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "C:\Users\mel\Anaconda3\envs\elm\lib\http\server.py", line 646, in __init__
    super().__init__(*args, **kwargs)
  File "C:\Users\mel\Anaconda3\envs\elm\lib\socketserver.py", line 720, in __init__
    self.handle()
  File "C:\Users\mel\Anaconda3\envs\elm\lib\http\server.py", line 426, in handle
    self.handle_one_request()
  File "C:\Users\mel\Anaconda3\envs\elm\lib\http\server.py", line 414, in handle_one_request
    method()
  File "C:\Users\mel\Anaconda3\envs\elm\lib\site-packages\xonsh\webconfig\main.py", line 93, in do_POST
    fname = insert_into_xonshrc(config)
  File "C:\Users\mel\Anaconda3\envs\elm\lib\site-packages\xonsh\webconfig\main.py", line 63, in insert_into_xonshrc         s = f.read()
  File "C:\Users\mel\Anaconda3\envs\elm\lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x90 in position 558: character maps to <undefined>
----------------------------------------
```

**Comment by melund** [comment]
@scopatz I think we should just merge this now. I will try to make some new PRs that fixes the windows issues I found.

**Pull request merged**